builder/docker/builder.go

```diff
@@ -120,46 +127,90 @@ func imageBuild(client provision.BuilderDockerClient, app provision.App, imageID
 		ProcfileRaw: procfileBuf.String(),
 		ImageID:     imageID,
 		Out:         evt,
-		CustomData:  customData,
+		CustomData:  tsuruYamlToCustomData(yaml),
 	})
 	if err != nil {
 		return "", err
 	}
 	return newImage, nil
 }
 
-func loadTsuruYaml(client provision.BuilderDockerClient, app provision.App, imageID string, evt *event.Event) (map[string]interface{}, error) {
+func loadTsuruYaml(client provision.BuilderDockerClient, app provision.App, imageID string, evt *event.Event) (*provision.TsuruYamlData, string, error) {
 	path := defaultArchivePath + "/current"
 	cmd := fmt.Sprintf("(cat %s/tsuru.yml || cat %s/tsuru.yaml || cat %s/app.yml || cat %s/app.yaml || true) 2>/dev/null", path, path, path, path)
 	var buf bytes.Buffer
-	err := runCommandInContainer(client, evt, imageID, cmd, app, &buf, nil)
+	containerID, err := runCommandInContainer(client, evt, imageID, cmd, app, &buf, nil)
 	if err != nil {
-		return nil, err
+		return nil, containerID, err
 	}
 	var tsuruYamlData provision.TsuruYamlData
 	err = yaml.Unmarshal(buf.Bytes(), &tsuruYamlData)
 	if err != nil {
-		return nil, err
+		return nil, containerID, err
 	}
-	customData := map[string]interface{}{
-		"healthcheck": tsuruYamlData.Healthcheck,
+	return &tsuruYamlData, containerID, err
+}
+
+func tsuruYamlToCustomData(yaml *provision.TsuruYamlData) map[string]interface{} {
+	if yaml == nil {
+		return nil
+	}
+
+	return map[string]interface{}{
+		"healthcheck": yaml.Healthcheck,
+		"hooks": map[string]interface{}{
+			"build": yaml.Hooks.Build,
+			"restart": map[string]interface{}{
+				"before": yaml.Hooks.Restart.Before,
+				"after":  yaml.Hooks.Restart.After,
+			},
+		},
 	}
-	return customData, err
 }
 
-func runCommandInContainer(client provision.BuilderDockerClient, evt *event.Event, image string, command string, app provision.App, stdout, stderr io.Writer) error {
+func runBuildHooks(client provision.BuilderDockerClient, app provision.App, imageID string, evt *event.Event, tsuruYamlData *provision.TsuruYamlData) (string, error) {
+	if tsuruYamlData == nil {
+		return imageID, nil
+	}
+	if len(tsuruYamlData.Hooks.Build) == 0 {
+		return "", nil
+	}
+
+	cmd := strings.Join(tsuruYamlData.Hooks.Build, " && ")
+	fmt.Fprintln(evt, "---- Running build hooks ----")
+	fmt.Fprintf(evt, " ---> Running %q\n", cmd)
+	containerID, err := runCommandInContainer(client, evt, imageID, cmd, app, evt, nil)
+	if err != nil {
+		return "", err
+	}
+
+	repo, tag := splitImageName(imageID)
+	opts := docker.CommitContainerOptions{
+		Container:  containerID,
+		Repository: repo,
```

Comment:
I think the only reason why `Repository` and `Tag` is needed here is because of a bug in docker-cluster, as it isn't storing information about the image if they are not present. Let's leave it like this for now but we may be able to simplify this in the future.